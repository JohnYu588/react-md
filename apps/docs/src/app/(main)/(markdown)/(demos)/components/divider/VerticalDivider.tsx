import { AppBar } from "@react-md/core/app-bar/AppBar";
import { AppBarTitle } from "@react-md/core/app-bar/AppBarTitle";
import { Divider } from "@react-md/core/divider/Divider";
import FavoriteIcon from "@react-md/material-icons/FavoriteIcon";
import { type ReactElement } from "react";

import styles from "./VerticalDivider.module.scss";

export default function VerticalDividerExample(): ReactElement {
  return (
    <AppBar theme="surface" className={styles.container}>
      <FavoriteIcon />
      <Divider vertical />
      <AppBarTitle>Title</AppBarTitle>
    </AppBar>
  );
}
