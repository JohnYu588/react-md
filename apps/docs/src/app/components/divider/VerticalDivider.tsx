import { AppBar, AppBarTitle, Divider } from "@react-md/core";
import FavoriteIcon from "@react-md/material-icons/FavoriteIcon";
import { type ReactElement } from "react";
import styles from "./VerticalDivider.module.scss";

export default function VerticalDividerExample(): ReactElement {
  return (
    <AppBar theme="surface" className={styles.container}>
      <FavoriteIcon />
      <Divider vertical />
      <AppBarTitle>Title</AppBarTitle>
    </AppBar>
  );
}
