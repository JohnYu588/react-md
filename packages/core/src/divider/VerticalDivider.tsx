import type { HTMLAttributes } from "react";
import { forwardRef } from "react";
import { Divider } from "./Divider";
import type { VerticalDividerHookOptions } from "./useVerticalDividerHeight";
import { useVerticalDividerHeight } from "./useVerticalDividerHeight";

export interface VerticalDividerProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * @see {@link VerticalDividerHookOptions.maxHeight}
   * @defaultValue `1`
   */
  maxHeight?: VerticalDividerHookOptions<HTMLDivElement>["maxHeight"];
}

/**
 * **Client Component**
 *
 * This component is used to create a vertical divider based on a parent
 * element's height. This is really only needed when the parent element **has no
 * defined height**.  If there is a defined height, this component is not worth
 * much as the height can be computed in css as normal. This really just fixes
 * the issue that the height would be set to `auto` (which computes to 0 most of
 * the time) when it is not set on a parent element.
 *
 * @example
 * Server Component Implementation
 * ```tsx
 * import type { ReactElement } from "react";
 * import { Divider } from "@react-md/core";
 *
 * function Example(): ReactElement {
 *   // or use a custom className
 *   return <Divider vertical style={{ height: "80%" }} />;
 * }
 * ```
 */
export const VerticalDivider = forwardRef<HTMLDivElement, VerticalDividerProps>(
  function VerticalDivider(props, ref) {
    const { style, maxHeight = 1, ...remaining } = props;

    const heightProps = useVerticalDividerHeight({
      ref,
      style,
      maxHeight,
    });

    return <Divider {...remaining} {...heightProps} vertical />;
  }
);
