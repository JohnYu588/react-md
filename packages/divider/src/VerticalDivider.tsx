import type { HTMLAttributes } from "react";
import { forwardRef } from "react";

import { Divider } from "./Divider";
import type { VerticalDividerHookOptions } from "./useVerticalDividerHeight";
import { useVerticalDividerHeight } from "./useVerticalDividerHeight";

export interface VerticalDividerProps extends HTMLAttributes<HTMLDivElement> {
  /**
   * @see {@link VerticalDividerHookOptions.maxHeight}
   * @defaultValue `1`
   */
  maxHeight?: VerticalDividerHookOptions<HTMLDivElement>["maxHeight"];
}

/**
 * This component is used to create a vertical divider based on a parent
 * element's height. This is really only needed when the parent element **has no
 * defined height**.  If there is a defined height, this component is not worth
 * much as the height can be computed in css as normal. This really just fixes
 * the issue that the height would be set to `auto` (which computes to 0 most of
 * the time) when it is not set on a parent element.
 */
export const VerticalDivider = forwardRef<HTMLDivElement, VerticalDividerProps>(
  function VerticalDivider(props, ref) {
    const { style, maxHeight = 1, ...remaining } = props;

    const heightProps = useVerticalDividerHeight({
      ref,
      style,
      maxHeight,
    });

    return <Divider {...remaining} {...heightProps} vertical />;
  }
);
