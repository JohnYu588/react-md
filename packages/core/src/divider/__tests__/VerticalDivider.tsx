import { describe, expect, it } from "@jest/globals";
import { createRef } from "react";
import { render } from "../../test-utils/index.js";

import { VerticalDivider } from "../VerticalDivider.js";

describe("VerticalDivider", () => {
  it("should apply the correct styling, HTML attributes, and allow a ref", () => {
    const ref = createRef<HTMLDivElement>();

    const { getByRole, rerender } = render(<VerticalDivider ref={ref} />);
    const divider = getByRole("separator");
    expect(ref.current).toBeInstanceOf(HTMLDivElement);
    expect(ref.current).toBe(divider);
    expect(divider).toMatchSnapshot();

    rerender(
      <VerticalDivider
        ref={ref}
        style={{ color: "blue" }}
        className="custom-class-name"
      />
    );
    expect(divider).toMatchSnapshot();

    rerender(<VerticalDivider ref={ref} maxHeight={0.8} />);
    expect(divider).toMatchSnapshot();

    rerender(<VerticalDivider ref={ref} maxHeight={300} />);
    expect(divider).toMatchSnapshot();

    rerender(<VerticalDivider ref={ref} maxHeight={-10} />);
    expect(divider).toMatchSnapshot();
  });
});
