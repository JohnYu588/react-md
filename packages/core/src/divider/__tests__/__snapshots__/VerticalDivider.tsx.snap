// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalDivider should apply the correct styling, HTML attributes, and allow a ref 1`] = `
<div
  class="rmd-divider rmd-divider--vertical"
  role="separator"
  style="height: 0px;"
/>
`;

exports[`VerticalDivider should apply the correct styling, HTML attributes, and allow a ref 2`] = `
<div
  class="rmd-divider rmd-divider--vertical custom-class-name"
  role="separator"
  style="height: 0px;"
/>
`;

exports[`VerticalDivider should apply the correct styling, HTML attributes, and allow a ref 3`] = `
<div
  class="rmd-divider rmd-divider--vertical"
  role="separator"
  style="height: 0px;"
/>
`;

exports[`VerticalDivider should apply the correct styling, HTML attributes, and allow a ref 4`] = `
<div
  class="rmd-divider rmd-divider--vertical"
  role="separator"
  style="height: 0px;"
/>
`;

exports[`VerticalDivider should apply the correct styling, HTML attributes, and allow a ref 5`] = `
<div
  class="rmd-divider rmd-divider--vertical"
  role="separator"
  style=""
/>
`;
